deeppavlov/models/elmo/elmo.py

```diff
@@ -0,0 +1,456 @@
+                for model in self.models
+            }
+        else:
+            feed_dict = {
+                model.tokens_characters:
+                    np.zeros([batch_size, unroll_steps, max_chars],
+                             dtype=np.int32)
+                for model in self.models
+            }
+
+        if self.options['bidirectional']:
+            if not char_inputs:
+                feed_dict.update({
+                    model.token_ids_reverse:
+                        np.zeros([batch_size, unroll_steps], dtype=np.int64)
+                    for model in self.models
+                })
+            else:
+                feed_dict.update({
+                    model.tokens_characters_reverse:
+                        np.zeros([batch_size, unroll_steps, max_chars],
+                                 dtype=np.int32)
+                    for model in self.models
+                })
+
+        init_state_values = self.sess.run(init_state_tensors, feed_dict=feed_dict)
+        return init_state_values, init_state_tensors, final_state_tensors
+
+    def _fill_feed_dict(self, 
+                        char_ids_batches, 
+                        reversed_char_ids_batches, 
+                        token_ids_batches = None, 
+                        reversed_token_ids_batches = None):
+        # init state tensors
+        feed_dict = {t: v for t, v in zip(self.init_state_tensors, self.init_state_values)}
+
+        for k, model in enumerate(self.models):
+            start = k * self.options['batch_size']
+            end = (k + 1) * self.options['batch_size']
+
+            # character inputs
+            char_ids = char_ids_batches[start:end]  # get char_ids
+
+            feed_dict[model.tokens_characters] = char_ids
+
+            if self.options['bidirectional']:
+                feed_dict[model.tokens_characters_reverse] = \
+                    reversed_char_ids_batches[start:end]  # get tokens_characters_reverse
+
+            if token_ids_batches is not None:
+                feed_dict[model.next_token_id] = token_ids_batches[start:end]  # get next_token_id
+                if self.options['bidirectional']:
+                    feed_dict[model.next_token_id_reverse] = \
+                        reversed_token_ids_batches[start:end]  # get next_token_id_reverse
+
+        return feed_dict
+
+    def __call__(self, *args, **kwargs) -> List[float]:
+        if len(args) != 2:
+            return []
+        char_ids_batches, reversed_char_ids_batches = args[0]
+        token_ids_batches, reversed_token_ids_batches = args[1]
+
+        feed_dict = self._fill_feed_dict(char_ids_batches, reversed_char_ids_batches, token_ids_batches, 
+                                         reversed_token_ids_batches)
+
+        with self.graph.as_default():
+            ret = self.sess.run([self.loss] + self.final_state_tensors, feed_dict)
+
+        self.init_state_values = ret[1:]
+        loss = ret[0]
+        return [loss]
```

Comment:
In `__call__` method you return `[loss]`, and in `train_on_batch` just `loss`. Is it okay?

Reply:
Yes for current version without infer mode. Now returned loss (from `__call__`) is used for validation score. train_on_batch is used for a real-time loss trade monitoring.